Usbhs

conorpp opened the pull request:
Add USB high speed peripheral support.

Summary of changes:
- Add USB1 peripheral enable/disable
- Change usbd to accept a Usb trait, which both USB0 and USB1 peripherals now implement
- Edit endpoint registers to use different offsets for `nbytes` and `addroff`, depending on if using FS periperal or HS peripheral.
- Update usb example to use either HS or FS.
- Use [new pac release](https://github.com/nickray/lpc55-pac/pull/6) which doesn't exist yet :)

nickray commented:
Looking forward to putting this through its paces!

nickray commented:
I published the PAC, renamed to just `lpc55-pac`, and version bumped to v0.1.0 for a bit of posterity.

nickray commented on src/traits/usb.rs:
```diff
@@ -0,0 +1,10 @@
+    FullSpeed,
+    HighSpeed,
+}
+pub trait Usb <State>: Deref<Target = crate::raw::usb1::RegisterBlock> + Sync  {
```
I suggest setting `UsbSpeed` as associated constant. So you'd have:

```
pub trait Usb <State>: Deref<Target = crate::raw::usb1::RegisterBlock> + Sync  {
    const SPEED: UsbSpeed;

    fn get_speed(&self) -> Self:::SPEED;
}

The getter might not even be necessary, as the associated constants are public.

nickray commented on src/drivers/usbd/endpoint_registers.rs:
```diff
@@ -534,23 +537,37 @@ pub mod epr {
+        #[doc = "Bits 16:25 - Endpoint buffer NBytes while in full speed operation, or bits 11:25 for high speed operation."]
+        #[inline]
+        pub fn nbytes(&self, speed: UsbSpeed) -> NBYTESR {
+            let bits = match speed {
```
Here and elsewhere constants are repeated. Also maybe the compiler optimizes this once `speed` is an associated constant, but it would be clearer to have `From<UsbSpeed> for `Whatever` so you can remove the `match` blocks explicitly.

conorpp replied:
The match blocks would just be moved to the `From<UsbSpeed> for Whatever` though?

conorpp replied:
I think i understand now

nickray commented on src/drivers/usbd.rs:
```diff
@@ -206,14 +218,14 @@ impl usb_device::bus::UsbBus for UsbBus {
             unsafe {
                 // lower part is stored in endpoint registers
-                let databufstart = (constants::EP_MEM_ADDR >> 22) as u16;
+                let databufstart = constants::EP_MEM_ADDR as u32;
```
I think this is kinda a bug in the UM/PAC. For HS there's no flexibility, for USB you can only change the upper bits. Right now we use USB RAM anyway so I guess it's a moot point.

conorpp commented on src/drivers/usbd.rs:
```diff
@@ -150,7 +162,7 @@ impl usb_device::bus::UsbBus for UsbBus {
                     let buffer = self.ep_allocator.allocate_buffer(size as _)?;
                     ep.set_out_buf(buffer);
-                    debug_assert!(ep.is_out_buf_set());
+                    assert!(ep.is_out_buf_set());
```
My mistake, didn't mean to commit those changes

conorpp commented:
Addressed comments:
* fixed clock error check
* replaced assert with debug_assert
* feature gated the number of endpoints (5 or 6)
* use associative constant for UsbSpeed and only one match block per register field

nickray commented:
Cool, I'm getting somewhat unreliable enumeration for the HS example on Linux, but FS still works reliably so let's merge and take it from there.

In HS enumeration in any case I see:
```
[556550.760992] usb 1-2.1: config 1 interface 0 altsetting 0 endpoint 0x81 has an invalid bInterval 255, changing to 11
[556550.760996] usb 1-2.1: config 1 interface 1 altsetting 0 bulk endpoint 0x82 has invalid maxpacket 8
[556550.760999] usb 1-2.1: config 1 interface 1 altsetting 0 bulk endpoint 0x2 has invalid maxpacket 8
```
